activesupport/lib/active_support/actionable_error.rb

```diff
@@ -9,12 +9,99 @@ module ActiveSupport
   module ActionableError
     extend Concern
 
-    class NonActionable < StandardError; end
+    class NonActionable < StandardError; end # :nodoc:
+
+    Trigger = Struct.new(:actionable, :condition) do # :nodoc:
+      def act_on(error)
+        raise actionable if condition&.call(error)
+      end
+    end
+
+    class Definition # :nodoc:
+      def initialize(&block)
+        instance_exec(self, &block)
+      end
+
+      def build
+        actionable = self
+
+        Class.new(StandardError) do
+          include ActiveSupport::ActionableError
+
+          define_method :initialize do |message = nil|
+            super(message || actionable.message)
+          end
+
+          actionable.actions.each do |(name, block)|
+            action(name, &block)
+          end
+
+          trigger(actionable.trigger) unless actionable.trigger.empty?
```

Comment:
Does it make sense to have one of these without a trigger?

Reply:
Actuall, maybe no if you're defining it with the DSL.